### ISSUE TITLE
clang-analyzer-security.ArrayBound and clang-analyzer-security.ArrayBound false positives

### ISSUE THREAD

**Issue by nick-potenski** [llvm/llvm-project#222102]
`clang-analyzer-core.UndefinedBinaryOperatorResult` and `clang-analyzer-security.ArrayBound` both report a false-positive "garbage value" / out-of-bounds read on a `malloc`'d array that is fully initialized by a preceding `for` loop over the same bound. The false positive only appears when the pointer supplying that bound is also passed to an unrelated opaque/external function call between the initializing loop and the later read of the array. Removing that single call argument makes both warnings disappear<strike>, even though the argument cannot affect the array's contents</strike>.  The analyzer's own path notes confirm it walked the first loop (`for (i = 0; i < info->count; i++) done[i] = 0;`) to completion, correctly reasoning `i >= info->count` at loop exit — i.e. it did model `done[]` as fully initialized — yet in the final loop it still treats `done[i]` as holding a garbage/uninitialized value.

I've reproduced this locally on v20.1.8 with the following code as well as on Compiler Explorer ([here](https://godbolt.org/z/8xeWYefq7)), which uses v24.0.0 as of today.

**Minimal reproducer**
```c
#include <stdlib.h>

typedef struct { int count; } info_type;

extern void log_impl(const char *fmt, ...);
extern int process_item(int index);

int process_all(info_type const *info) {
  int *done = (int *)malloc(sizeof(int) * info->count);
  int status = 0;
  int i;

  for (i = 0; i < info->count; i++) {
    done[i] = 0;
  }

  status = process_item(0);
  log_impl("processing, info: %p", info); /* removing this call silences both warnings */

  if (status != 0) {
    for (i = 0; i < info->count; i++) {
      if (done[i] == 1) { /* false-positive garbage-value read reported here */
      }
    }
  }

  free(done);
  return status;
}
```
yields:
```
<source>:22:11: warning: Out of bound access to memory after the end of the heap area [clang-analyzer-security.ArrayBound]
   22 |       if (done[i] == 1) { /* false-positive garbage-value read reported here */
      |           ^~~~~~~
<source>:13:15: note: Assuming 'i' is < field 'count'
   13 |   for (i = 0; i < info->count; i++) {
      |               ^~~~~~~~~~~~~~~
<source>:13:3: note: Loop condition is true.  Entering loop body
   13 |   for (i = 0; i < info->count; i++) {
      |   ^
<source>:13:15: note: Assuming 'i' is >= field 'count'
   13 |   for (i = 0; i < info->count; i++) {
      |               ^~~~~~~~~~~~~~~
<source>:13:3: note: Loop condition is false. Execution continues on line 17
   13 |   for (i = 0; i < info->count; i++) {
      |   ^
<source>:20:7: note: Assuming 'status' is not equal to 0
   20 |   if (status != 0) {
      |       ^~~~~~~~~~~
<source>:20:3: note: Taking true branch
   20 |   if (status != 0) {
      |   ^
<source>:21:17: note: Assuming 'i' is < field 'count'
   21 |     for (i = 0; i < info->count; i++) {
      |                 ^~~~~~~~~~~~~~~
<source>:21:5: note: Loop condition is true.  Entering loop body
   21 |     for (i = 0; i < info->count; i++) {
      |     ^
<source>:22:7: note: Taking false branch
   22 |       if (done[i] == 1) { /* false-positive garbage-value read reported here */
      |       ^
<source>:21:17: note: Assuming 'i' is < field 'count'
   21 |     for (i = 0; i < info->count; i++) {
      |                 ^~~~~~~~~~~~~~~
<source>:21:5: note: Loop condition is true.  Entering loop body
   21 |     for (i = 0; i < info->count; i++) {
      |     ^
<source>:22:11: note: Access of 'int' element in the heap area at index 1
   22 |       if (done[i] == 1) { /* false-positive garbage-value read reported here */
      |           ^~~~~~~
<source>:22:19: warning: The left operand of '==' is a garbage value [clang-analyzer-core.UndefinedBinaryOperatorResult]
   22 |       if (done[i] == 1) { /* false-positive garbage-value read reported here */
      |           ~~~~~~~ ^
<source>:9:22: note: Storing uninitialized value
    9 |   int *done = (int *)malloc(sizeof(int) * info->count);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:13:15: note: Assuming 'i' is >= field 'count'
   13 |   for (i = 0; i < info->count; i++) {
      |               ^~~~~~~~~~~~~~~
<source>:13:3: note: Loop condition is false. Execution continues on line 17
   13 |   for (i = 0; i < info->count; i++) {
      |   ^
<source>:20:7: note: Assuming 'status' is not equal to 0
   20 |   if (status != 0) {
      |       ^~~~~~~~~~~
<source>:20:3: note: Taking true branch
   20 |   if (status != 0) {
      |   ^
<source>:21:10: note: The value 0 is assigned to 'i'
   21 |     for (i = 0; i < info->count; i++) {
      |          ^~~~~
<source>:21:17: note: Assuming 'i' is < field 'count'
   21 |     for (i = 0; i < info->count; i++) {
      |                 ^~~~~~~~~~~~~~~
<source>:21:5: note: Loop condition is true.  Entering loop body
   21 |     for (i = 0; i < info->count; i++) {
      |     ^
<source>:22:19: note: The left operand of '==' is a garbage value
   22 |       if (done[i] == 1) { /* false-positive garbage-value read reported here */
      |           ~~~~~~~ ^
```

**Comment by NagyDonat** [comment]
> The false positive only appears when the pointer supplying that bound is also passed to an unrelated opaque/external function call between the initializing loop and the later read of the array. Removing that single call argument makes both warnings disappear, even though the argument cannot affect the array's contents. 

The opaque function call cannot affect the array's contents, but it could easily change the value of `info->count`. To represent this, at the opaque call the analyzer forgets its knowledge about the value in `info->count` (to represent that it may have been changed), so it finds theoretically possible paths like "`info->count` is 0 at first, the initialization loop does nothing, `log_impl` updates `info->count` to 1, the second loop accesses `done[0]`" – where we do have the errors.

These "we assume one thing, variable is invalidated, we assume another thing about its value" false positives are unfortunately widespread in the analyzer, well-known by the developers, and difficult to fix. There is a draft PR https://github.com/llvm/llvm-project/pull/207155 that introduces some tools for fixing these kinds of issues – perhaps it could lead to resolving this issue.

I thought that we had an umbrella issue for this kind of false positives, but I was unable to find one. If there isn't one, we should probably create one and use it to describe the plans and challenges in a single central location.

**Comment by nick-potenski** [comment]
> The opaque function call cannot affect the array's contents, but it could easily change the value of info->count. 

Ya, sorry, that was poorly worded.  But info is defined as a pointer to const, so even if it was passed to an opaque function I wouldn't expect the analyzer to assume its contents are changed, unless const was cast away.  Does the analyzer just assume the worst in that case or is there something to do with the variadic argument that affects the constness?

**Comment by NagyDonat** [comment]
> But info is defined as a pointer to const,

Oh, good point, I didn't notice that detail 😅 

> so even if it was passed to an opaque function I wouldn't expect the analyzer to assume its contents are changed, unless const was cast away. Does the analyzer just assume the worst in that case or is there something to do with the variadic argument that affects the constness?

The invalidation is done by the method `CallEvent::invalidateRegions`. This method has logic which ensures that the contents of the area pointed-to by a pointer argument is not invalidated if the type of the corresponding parameter in the function declaration is "pointer to const".

If the function isn't variadic, this produces correct behavior, because if the argument is pointer-to-const, then the corresponding parameter also must be pointer-to-const (otherwise the code doesn't compile) so it is not invalidated. (But it also covers the case when the pointee of the argument is non-const, but it is passed to a pointer-to-const parameter and we still don't invalidate.)

However, if the function is variadic, then the pointer-to-const argument has no correspondig parameter in the declaration, so it isn't added to the "don't invalidate the pointee" list and it is (incorrectly) hit by the invalidation.

This is a bug in `CallEvent::invalidateRegions()`, and it would be relatively easy to resolve it by inserting (the indices of) the variadic arguments that have pointer-to-const types into the set `PreserveArgs`. (Note that there are _three_ conflicting conventions for indexing the argument/parameters of a `CallEvent`, see PR https://github.com/llvm/llvm-project/pull/221252 by @benedekaibas which tries to clarify their relationships.)

By the way, @benedekaibas do you happen to be interested in creating a patch that fixes this issue? This can be fixed with a small patch (10-15 lines) and IMO the chief difficulty is understanding the difference between the index kinds, and you already have that.

**Comment by benedekaibas** [comment]
> By the way, [@benedekaibas](https://github.com/benedekaibas) do you happen to be interested in creating a patch that fixes this issue? This can be fixed with a small patch (10-15 lines) and IMO the chief difficulty is understanding the difference between the index kinds, and you already have that.

Thank you for mentioning me! Yes, definitely!!! I will look into it and open a PR for it. Hmm maybe it can be stacked on the current PRs (#221252). I will look into it and use the best judgement to make the fix. @NagyDonat thanks for mentioning it, I greatly appreciate it!

**Comment by NagyDonat** [comment]
> I will look into it and open a PR for it.

Thank you very much!
